examples/hybrid_plugin_system.ipynb

```diff
@@ -0,0 +1,391 @@
+    "\n",
+    "try:\n",
+    "    import openassetio\n",
+    "    import openassetio_mediacreation\n",
+    "except ImportError:\n",
+    "    print(\n",
+    "        \"This notebook requires the packages listed in `resources/requirements.txt` to be installed\")\n",
+    "    raise\n",
+    "\n",
+    "from resources import helpers\n",
+    "\n",
+    "from openassetio.hostApi import HostInterface, ManagerFactory\n",
+    "from openassetio.log import LoggerInterface\n",
+    "\n",
+    "\n",
+    "class NotebookHostInterface(HostInterface):\n",
+    "    def identifier(self):\n",
+    "        return \"org.jupyter.notebook\"\n",
+    "\n",
+    "    def displayName(self):\n",
+    "        return \"Jupyter Notebook\"\n",
+    "    \n",
+    "    \n",
+    "class NullLogger(LoggerInterface):\n",
+    "    def log(self, _severity, _message):\n",
+    "        pass\n",
+    "    \n",
+    "\n",
+    "host_interface = NotebookHostInterface()\n",
+    "\n",
+    "logger = NullLogger()"
+   ],
+   "id": "db2739c3f0a96d70",
+   "outputs": [],
+   "execution_count": 8
+  },
+  {
+   "metadata": {},
+   "cell_type": "markdown",
+   "source": [
+    "\n",
+    "### The two example plugins\n",
+    "\n",
+    "In order to illustrate the hybrid plugin system, we'll make use of two ready-made example plugins - the \"Basic Asset Libary\" (aka BAL), a pure Python manager/plugin, and the \"Simple C++ Manager\" (aka SimpleCppManager), a pure C++ manager/plugin. \n",
+    "\n",
+    "BAL should be installed into the Python environment of this notebook (see `resources/requirements.txt`), and so will be trivially discoverable by OpenAssetIO.\n",
+    "\n",
+    "SimpleCppManager is more complex, and must be built with a compiler toolchain compatible with the OpenAssetIO libraries in the Python environment of this notebook. See `resources/hybrid_plugin_system/SimpleCppManager/README.md` for more details. We assume it is installed into `resources/hybrid_plugin_system/SimpleCppManager`, and will be discovered by adding this location to the standard `OPENASSETIO_PLUGIN_PATH` environment variable."
+   ],
+   "id": "791f52d06eea9aea"
+  },
+  {
+   "metadata": {
+    "ExecuteTime": {
+     "end_time": "2024-08-30T18:25:10.551282Z",
+     "start_time": "2024-08-30T18:25:10.546996Z"
+    }
+   },
+   "cell_type": "code",
+   "source": [
+    "os.environ[\"OPENASSETIO_PLUGIN_PATH\"] = os.path.join(\n",
+    "    \"resources\", \"hybrid_plugin_system\", \"SimpleCppManager\")"
+   ],
+   "id": "e3a94c5be66ad34e",
+   "outputs": [],
+   "execution_count": 9
+  },
+  {
+   "metadata": {},
+   "cell_type": "markdown",
+   "source": [
+    "\n",
+    "\n",
+    "These two plugins each advertise a different unique identifier, so will not work with the hybrid plugin system out of the box. Luckily, they are both designed to be used in testing situations and are fully configurable. We can control the identifier that they will advertise using environment variables:\n"
```

Comment:
[fix] Is there scope to make a toy code base with python/c++ manager plugins that already shares an identifier, and name. I totally get where this is coming from, I wonder though if I was new to all this it would introduce concepts that would muddy the waters, that we don't really intend to be every day practice.
- The idea of a manager having a configurable identifier.
- Combining two logically separate plugins through this mechanism.
- That its ok for two parts of the implementation to have different responses for the same method. (iirc, it is critical that you _are_ coherent between the two and don't have differences, as say managmentPolicy wrtitten in Python affects how C++ would be called)?.

Reply:
Yeah, I was uneasy about this too. It comes from the notebook's secondary purpose of being an e2e test, and trying to wrap the documentation around that.

I did wonder about doing something similar to the (currently very WIP) UI delegation notebook (https://github.com/OpenAssetIO/OpenAssetIO-MediaCreation/pull/100), where I've made a `NotebookPluginSystem` to use inline just for that notebook. So here, we could create two `NotebookPluginSystem`s and combine them with the hybrid plugin system.

But then BAL, SimpleCppManager, and in-particular the C++ plugin system, wouldn't get exercised.

I did at least explicitly call out that the two implementations should have the same response in the real world. Not sure how to illustrate/test priority ordering without the implementations giving different responses. Perhaps that bit should be cut if it causes confusion (there is already lower-level test coverage).

Reply:
Ahh that makes sense - so to play deviils advocate - I guess the Q is - what is most important? The e2e case, or orienting people?

Reply:
I tried a little restructuring to de-emphasize the boilerplate in 1113765e - does that help?

Reply:
I guess its really the BAL/SimpleCppManager thing that feels like it might throw people. 

It would be really useful for people to see how you might structure your code to write your own hybrid plugin too.  If you could have a really really basic minimal manager that only does resolve in C++ and everything else in python (only supporting say, locatable content trait, no relations or anything) it feels like it would be orders of magnitude more useful for getting people started. It could be the starting point for a hybrid version of [the template](https://github.com/OpenAssetIO/Template-OpenAssetIO-Manager-Python) too.

Hybrid managers are such a valuable feature of the API, that making "getting going with a hybrid plugin" as easy as possible would really help sell that befit and highlight the reason for adoption OpenAssetIO vs others.

Reply:
I think I agree in principle that it would be more instructive to have a dedicated example hybrid plugin, rather than sellotaping BAL and SimpleCppManager together.

But given the work involved in doing this, my question is, does this notebook really need this to be useful, or can this notebook be merged and a follow-up issue created to make a cleaner example?

Reply:
Maybe the question to ask is - what is the overhead caused by confusion of new-comers?

Perhaps its really a case of what is motivating this PR? Is it to provide clear information to users, or to make sure there is an integration test of the new functionality?

Reply:
Do you think its really that confusing?

I think the primary motivation at this stage of the process is indeed to make good documentation, other aspects are secondary.

If this notebook is likely to add rather than remove confusion then I'd agree it needs re-thinking.  But if the notebook does its job as documentation and its a case of optimising the notebook, then that could come later.

Reply:
I might not be the best reference point - maybe one to ask on the slack channel?

I guess I was thinking as to how people often struggle to see how BAL fits into the ecosystem on its own, and it often gets promoted beyond its station. I was worried that coupling this with another, differently named project, and introducing the idea of changing a managers identifier might introduce even more complexity.

How much effort do you think it would be to make the most basic hybrid `SampleHybridManager` project, that _only_ supports `resolve` in C++, and the most minimal, minimal implementation of just the bare essentials on the Python side? `resolve` could just echo the ref back without a scheme or similar.   This would serve both as a staring point for some later template project like Elliot's Python one, and would also iron out any gremlins in this idea of minimal implementations in some languages (guessing BAL and SimpleCppManager are valid on their own, which might mask some rough edges on capability splicing that hadn't surfaced otherwise).

2p!

Reply:
> The reason for leaning towards deferring improvements a bit is that there may be other DCC/Asset Manager examples that fall out in the wash from work we might tackle in the coming months, and those may represent better bang for buck.

2p: We always held the "hybrid plugin system" as one of the unique value-adds it the API can bring, so anything that helps people get started developing a hybrid plugin would be a real boon to adoption (our USD Ar plugin is a hybrid).

Reply:
OK, had a go at using a bespoke hybrid plugin just for this notebook in cf57644e

Also added a note about using the `Context` to communicate between plugins.

Reply:
This is fantastic @feltech  🙏  - so awesome to see how little code that is.  How was making it - did it highlight anything interesting? I could see someone wanting to get going with the API and a hybrid plugin could really easily use this as a starting point. It feels like its orders of magnitude easier to grok as it doesn't contain any of the fluff BAL has in order to fulfill its role.